flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/stream/MiniBatchIntervalInferRule.scala

```diff
@@ -120,9 +116,22 @@ class MiniBatchIntervalInferRule extends RelOptRule(
     }
   }
 
-  private def isTableSourceScan(node: RelNode): Boolean = node match {
-    case _: StreamExecDataStreamScan | _: StreamExecTableSourceScan => true
-    case _ => false
+  private def shouldAppendMiniBatchAssignerNode(node: RelNode): Boolean = {
+    val mode = node.getTraitSet
+      .getTrait(MiniBatchIntervalTraitDef.INSTANCE)
+      .getMiniBatchInterval
+      .mode
+    node match {
+      case _: StreamExecDataStreamScan | _: StreamExecTableSourceScan =>
+        // append minibatch node if the mode is not NONE and reach a source leaf node
+        mode == MiniBatchMode.RowTime || mode == MiniBatchMode.ProcTime
+      case _: StreamExecWatermarkAssigner  =>
```

Comment:
IIUC, for the case that there is only group aggregation with a **redundant watermark** defined in the sql job, minibatch will not worked even when the user  set `table.exec.mini-batch.enabled = true`, which confuses users as this is transparent to them. I'm wondering if some checks can be added to alert user?

Reply:
Good catch! We can still append mini-batch node after watermark node if it is proctime mode, because proctime minibatch assginer will override upstream watermarks. I updated the code.